dex/fiatrates/oracle.go

```diff
@@ -0,0 +1,239 @@
+package fiatrates
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+const (
+	// FiatRateDataExpiry : Any data older than FiatRateDataExpiry will be
+	// discarded.
+	FiatRateDataExpiry = 60 * time.Minute
+
+	// averageRateRefreshInterval is how long it'll take before a fresh fiat
+	// average rate is calculated.
+	averageRateRefreshInterval = defaultRefreshInterval + time.Minute
+)
+
+// Oracle manages and retrieves fiat rate information from all enabled rate
+// sources.
+type Oracle struct {
+	sources  []*source
+	ratesMtx sync.RWMutex
+	rates    map[string]*FiatRateInfo
+}
+
+func NewFiatOracle(cfg Config) (*Oracle, error) {
+	fiatOracle := &Oracle{
+		rates:   make(map[string]*FiatRateInfo),
+		sources: fiatSources(cfg),
+	}
+
+	assets := strings.Split(cfg.Assets, ",")
+	for _, asset := range assets {
+		_, ok := dex.BipSymbolID(strings.ToLower(asset))
+		if !ok {
+			return nil, fmt.Errorf("unknown asset %s", asset)
+		}
+
+		// Initialize entry for this asset.
+		fiatOracle.rates[parseSymbol(asset)] = new(FiatRateInfo)
+	}
+
+	return fiatOracle, nil
+}
+
+// Rate returns the current fiat rate information for the provided symbol.
+// Returns zero if there are no rates for the provided symbol yet.
+func (o *Oracle) Rate(symbol string) float64 {
+	o.ratesMtx.Lock()
+	defer o.ratesMtx.Unlock()
+
+	symbol = parseSymbol(symbol)
+	rateInfo := o.rates[symbol]
+	hasRateInfo := rateInfo != nil
+	if hasRateInfo && time.Since(rateInfo.LastUpdate) < FiatRateDataExpiry && rateInfo.Rate > 0 {
+		return rateInfo.Rate
+	}
+
+	if !hasRateInfo {
+		// Initiate an entry for this asset. Data for this asset will be fetched
+		// in the next refresh cycle.
+		o.rates[symbol] = new(FiatRateInfo)
+	}
+
+	return 0
```

Comment:
I still think it would be best to short circuit here and add the new rate, even if it's only from one source.

Reply:
Okay, will add in my next push.